In [75]:
### Import all necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import warnings 
warnings.filterwarnings('ignore')
from sklearn.preprocessing import StandardScaler, MinMaxScaler ### Scaling
from sklearn.preprocessing import OneHotEncoder,LabelEncoder  ### Encoding

In [41]:
#Load the dataset
df=pd.read_csv('adult_with_headers (1).csv')

In [42]:
### Check the size of data
df.shape

(32561, 15)

In [43]:
### data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [76]:
### Statistical Summary
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [31]:
df.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


#### 1. EDA

In [32]:
### Duplicates
df.duplicated().sum()

np.int64(24)

In [33]:
### Drop the duplicates
df.drop_duplicates(inplace=True,ignore_index=True)

In [34]:
df.duplicated().sum()

np.int64(0)

In [39]:
### Missing values
df.isnull().sum().sum()

np.int64(0)

#### Scaling

In [48]:
### Select Numerical columns
num_cols = df.select_dtypes(include=['int64', 'float64']).columns
num_cols

Index(['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week'],
      dtype='object')


In [50]:
std_sca=StandardScaler()
min_max=MinMaxScaler()

In [52]:
###Standard Scaler
df_standard = df.copy()
df_standard[num_cols] = std_sca.fit_transform(df_standard[num_cols])
df_standard.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,0.030671,State-gov,-1.063611,Bachelors,1.134739,Never-married,Adm-clerical,Not-in-family,White,Male,0.148453,-0.21666,-0.035429,United-States,<=50K
1,0.837109,Self-emp-not-inc,-1.008707,Bachelors,1.134739,Married-civ-spouse,Exec-managerial,Husband,White,Male,-0.145920,-0.21666,-2.222153,United-States,<=50K
2,-0.042642,Private,0.245079,HS-grad,-0.420060,Divorced,Handlers-cleaners,Not-in-family,White,Male,-0.145920,-0.21666,-0.035429,United-States,<=50K
3,1.057047,Private,0.425801,11th,-1.197459,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,-0.145920,-0.21666,-0.035429,United-States,<=50K
4,-0.775768,Private,1.408176,Bachelors,1.134739,Married-civ-spouse,Prof-specialty,Wife,Black,Female,-0.145920,-0.21666,-0.035429,Cuba,<=50K


##### Standard Scaling

**Preferred when:**

* Numerical features have different ranges.
* The data is approximately normally distributed.
* Models such as Logistic Regression, SVM, KNN, PCA, and Linear Regression are used.

**why:**

* Standard Scaling changes the data so that it has a mean of 0 and a standard deviation of 1.
* It ensures that all numerical features are on a similar scale.
* This prevents features with larger values from having a greater influence on the model.
* As a result, many machine learning algorithms can learn more effectively and produce better results.

In [53]:
### min-max Scaler
df_minmax = df.copy()
df_minmax[num_cols] = min_max.fit_transform(df_minmax[num_cols])
df_minmax.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,0.301370,State-gov,0.044302,Bachelors,0.800000,Never-married,Adm-clerical,Not-in-family,White,Male,0.02174,0.0,0.397959,United-States,<=50K
1,0.452055,Self-emp-not-inc,0.048238,Bachelors,0.800000,Married-civ-spouse,Exec-managerial,Husband,White,Male,0.00000,0.0,0.122449,United-States,<=50K
2,0.287671,Private,0.138113,HS-grad,0.533333,Divorced,Handlers-cleaners,Not-in-family,White,Male,0.00000,0.0,0.397959,United-States,<=50K
3,0.493151,Private,0.151068,11th,0.400000,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0.00000,0.0,0.397959,United-States,<=50K
4,0.150685,Private,0.221488,Bachelors,0.800000,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0.00000,0.0,0.397959,Cuba,<=50K


##### Min-Max Scaling

**Preferred when:**

* Numerical features need to be in the range of 0 to 1.
* Neural Networks and other algorithms that work better with normalized data are used.
* The dataset has very few extreme outliers.

**Why:**

* Min-Max Scaling rescales all numerical features to a fixed range between 0 and 1.
* It keeps the relative differences between data points while making all features have the same scale.
* This helps models that require bounded input values train more efficiently.
* It is a good choice when the original data does not contain many extreme values.

#### 2. Encoding Techniques:

In [67]:
cat_col = [col for col in df.select_dtypes(include='object').columns if df[col].nunique() < 5] ### less than 5 categories
cat_col

['sex', 'income']

In [65]:
### One-Hot Encoder
one_hot=OneHotEncoder()
one_hot = pd.get_dummies(df, dtype='int',columns=cat_col, drop_first=True)
one_hot.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Male,income_ >50K
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,2174,0,40,United-States,1,0
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,0,0,13,United-States,1,0
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,0,0,40,United-States,1,0
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,0,0,40,United-States,1,0
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,0,0,40,Cuba,0,0


    One-Hot Encoding was applied to categorical variables with less than 5 categories. Each category was converted into a separate binary column with values of 0 and 1. This helps the model treat each category independently without assuming any relationship between them.

In [68]:
### Label Encoding
cat_col = [col for col in df.select_dtypes(include='object').columns if df[col].nunique() >5] ### Greater than 5 categories
cat_col

['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'native_country']

In [69]:
df_label = df.copy()
lab_en = LabelEncoder()
for col in cat_col:
    df_label[col] = lab_en.fit_transform(df_label[col])
    
df_label.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,7,77516,9,13,4,1,1,White,Male,2174,0,40,39,<=50K
1,50,6,83311,9,13,2,4,0,White,Male,0,0,13,39,<=50K
2,38,4,215646,11,9,0,6,1,White,Male,0,0,40,39,<=50K
3,53,4,234721,1,7,2,6,0,Black,Male,0,0,40,39,<=50K
4,28,4,338409,9,13,2,10,5,Black,Female,0,0,40,5,<=50K


    Label Encoding was applied to categorical variables with more than 5 categories. Each unique category was converted into a unique numerical value. This reduces the number of columns compared to One-Hot Encoding and makes the dataset more memory-efficient.

#### Pros and Cons of One-Hot Encoding
**Pros**
* Easy to understand and implement.
* Does not create any order between categories.
* Works well for categorical variables with a small number of categories.
* Suitable for many machine learning algorithms.
  
**Cons**
* Creates additional columns for each category.
* Increases the size of the dataset when there are many categories.
* Can increase memory usage and computation time.

#### Pros and Cons of Label Encoding

**Pros**
* Easy to apply.
* Does not increase the number of columns.
* Uses less memory than One-Hot Encoding.
* Suitable for categorical variables with many categories.
  
**Cons**
* Assigns numerical values to categories, which may create an artificial order.
* Some machine learning algorithms may incorrectly assume that larger numbers represent higher importance.
* May reduce model performance for nominal categorical variables.

#### 3. Feature Engineering:

In [70]:
### Age Group
df_feature = df.copy()

df_feature['Age_Group'] = pd.cut(
    df_feature['age'],
    bins=[0, 25, 40, 60, 100],
    labels=['Young', 'Adult', 'Middle_Aged', 'Senior']
)

df_feature.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income,Age_Group
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K,Adult
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K,Middle_Aged
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K,Adult
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K,Middle_Aged
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K,Adult


    Age_Group was created by grouping individuals based on their age. People in different age groups may have different income levels and work experience, making this feature useful for prediction.

In [71]:
### Total capital
df_feature['Total_Capital'] = df_feature['capital_gain'] - df_feature['capital_loss']

df_feature.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income,Age_Group,Total_Capital
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K,Adult,2174
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K,Middle_Aged,0
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K,Adult,0
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K,Middle_Aged,0
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K,Adult,0


    Total_Capital was created by combining capital_gain and capital_loss. It represents the overall capital value of an individual, which may have a stronger relationship with income than considering these two features separately.

In [72]:
num_cols = df_feature.select_dtypes(include=['int64', 'float64']).columns

df_feature[num_cols].skew()

age                0.558743
fnlwgt             1.446980
education_num     -0.311676
capital_gain      11.953848
capital_loss       4.594629
hours_per_week     0.227643
Total_Capital     11.871934
dtype: float64

#### Log Transformation
    The capital_gain feature was selected for log transformation because it is highly positively skewed (skewness = 11.95). Most observations have a value of 0, while a few observations have very large values.

    Applying log1p() reduces the skewness, decreases the impact of extreme values, and creates a more balanced distribution. This can improve the performance of machine learning models.

In [73]:
df_feature['capital_gain_log'] = np.log1p(df_feature['capital_gain'])
df_feature.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income,Age_Group,Total_Capital,capital_gain_log
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K,Adult,2174,7.684784
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K,Middle_Aged,0,0.000000
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K,Adult,0,0.000000
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K,Middle_Aged,0,0.000000
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K,Adult,0,0.000000
